include/ebpf_registry_helper.h

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <ntddk.h>
+
+#define REG_CREATE_FLAGS 0
+#define GUID_STRING_LENGTH 38 // not inlcuding the null terminator.
+
+typedef struct _ebpf_registry_key
+{
+    HANDLE handle;
+} ebpf_registry_key_t;
+
+static void
+close_registry_key(_In_ ebpf_registry_key_t* key)
+{
+    if (key->handle != NULL) {
+        ZwClose(key->handle);
+        key->handle = NULL;
+    }
+}
+
+_Success_(return == 0) static NTSTATUS
+    convert_guid_to_string(_In_ const GUID* guid, _Out_writes_all_(string_size) wchar_t* string, size_t string_size)
+{
+    UNICODE_STRING unicode_string;
+    NTSTATUS status = RtlStringFromGUID(guid, &unicode_string);
+    if (status != STATUS_SUCCESS) {
+        goto Exit;
+    }
+
+    if (string_size < GUID_STRING_LENGTH + 1) {
+        status = STATUS_BUFFER_TOO_SMALL;
+        goto Exit;
+    }
+
+    __analysis_assume(unicode_string.MaximumLength >= GUID_STRING_LENGTH * 2);
+    __analysis_assume(unicode_string.Buffer != NULL);
+    // Copy the buffer to the output string.
+    memcpy(string, unicode_string.Buffer, GUID_STRING_LENGTH * 2);
+    string[GUID_STRING_LENGTH] = L'\0';
+
+Exit:
+    return status;
+}
+
+static uint32_t
+write_registry_value_binary(
+    _In_ const ebpf_registry_key_t* key,
+    _In_ const wchar_t* value_name,
+    _In_reads_(value_size) uint8_t* value,
+    _In_ size_t value_size)
+{
+    UNICODE_STRING unicode_value_name;
+
+    RtlInitUnicodeString(&unicode_value_name, value_name);
+    return ZwSetValueKey(key->handle, &unicode_value_name, 0, REG_BINARY, value, (ULONG)value_size);
+}
+
+static uint32_t
+write_registry_value_ansi_string(
+    _In_ const ebpf_registry_key_t* key, _In_ const wchar_t* value_name, _In_z_ const char* value)
+{
+    NTSTATUS status;
+    UNICODE_STRING unicode_value;
+    UNICODE_STRING unicode_value_name;
+
+    ANSI_STRING ansi_string;
+    RtlInitAnsiString(&ansi_string, value);
+
+    status = RtlAnsiStringToUnicodeString(&unicode_value, &ansi_string, TRUE);
+    if (!NT_SUCCESS(status)) {
+        goto Exit;
+    }
+    RtlInitUnicodeString(&unicode_value_name, value_name);
+
+    status = ZwSetValueKey(key->handle, &unicode_value_name, 0, REG_SZ, unicode_value.Buffer, unicode_value.Length);
+    RtlFreeUnicodeString(&unicode_value);
+
+Exit:
+    return status;
+}
+
+static uint32_t
+write_registry_value_dword(_In_ const ebpf_registry_key_t* key, _In_z_ const wchar_t* value_name, uint32_t value)
+{
+    UNICODE_STRING unicode_name;
+    RtlInitUnicodeString(&unicode_name, value_name);
+    return ZwSetValueKey(key->handle, &unicode_name, 0, REG_DWORD, &value, sizeof(uint32_t));
+}
+
+static uint32_t
+create_registry_key(
```

Comment:
Several questions:
What is the need of these wrapper functions over registry if we already have the abstraction of store? Why cant the store implementation directly invoke registry APIs? And why is the library implemented in header instead of c/cpp and packaged into a library?

Reply:
1. store APIs are called from both user mode and kernel mode. In kernel, the extension drivers call these APIs. In user mode, export_program_info.exe tool uses these APIs to export the static program info. I wanted to keep as much code in the store APIs as possible common between user mode and kernel mode, hence wrappers over registry APIs.
2. Only store APIs are the kernel registry files are in header file. The user mode registry APIs are implemented in api_common.ib.
3. Agree that we can have a .lib for kernel mode which extension drivers will link to, to use the APIs. I just implemented them in the header file so an extension author will just need to include the header file.
4.